exp/runtime/internal/controllers/extension_controller_test.go

```diff
@@ -315,7 +313,33 @@ func TestExtensionReconciler_Reconcile(t *testing.T) {
 				}
 			}
 		}
-		g.Expect(env.CleanupAndWait(ctx, workingExtension1, brokenExtension)).To(Succeed())
+		g.Expect(env.CleanupAndWait(ctx, extensionWithDiscovery1, extensionWithDiscovery2)).To(Succeed())
+	})
+
+	t.Run("test removal of extension", func(t *testing.T) {
+		var result runtimev1.Extension
+
+		r := &Reconciler{
+			Client:    env.GetClient(),
+			APIReader: env.GetAPIReader(),
+			RuntimeClient: &fakeClient{
+				extensionStore: hookStore,
+			},
+		}
+
+		// Create and attempt to discover a working, discoverable extension. Expect no error.
+		g.Expect(env.CreateAndWait(ctx, workingExtension1.DeepCopy())).To(Succeed())
+		_, err := r.Reconcile(ctx, ctrl.Request{NamespacedName: namespacedName(workingExtension1)})
+		g.Expect(err).To(BeNil())
+		g.Expect(env.GetAPIReader().Get(ctx, namespacedName(workingExtension1), &result)).To(Succeed())
+
+		// Delete and attempt to discover a removal of the extension. Expect no error.
+		g.Expect(env.Delete(ctx, workingExtension1.DeepCopy())).To(Succeed())
```

Comment:
We might want to use `env.CleanupAndWait` instead of `env.Delete`. I have observed cases where using `env.Delete` causes race conditions in the tests and the reconciler might act before the cache is updated and could lead to test flakes.